extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddStatefulSetResourceDecorator.java

```diff
@@ -1,43 +1,80 @@
 
 package io.quarkus.kubernetes.deployment;
 
+import static io.quarkus.kubernetes.deployment.Constants.STATEFULSET;
+
 import java.util.HashMap;
+import java.util.List;
 
 import io.dekorate.kubernetes.decorator.ResourceProvidingDecorator;
+import io.dekorate.utils.Strings;
+import io.fabric8.kubernetes.api.model.Container;
 import io.fabric8.kubernetes.api.model.KubernetesListFluent;
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
 import io.fabric8.kubernetes.api.model.apps.StatefulSetBuilder;
+import io.fabric8.kubernetes.api.model.apps.StatefulSetFluent;
 
 public class AddStatefulSetResourceDecorator extends ResourceProvidingDecorator<KubernetesListFluent<?>> {
 
     private final String name;
     private final PlatformConfiguration config;
 
+    public AddStatefulSetResourceDecorator(String name, PlatformConfiguration config) {
+        this.name = name;
+        this.config = config;
+    }
+
+    @SuppressWarnings("deprecation")
     @Override
     public void visit(KubernetesListFluent<?> list) {
-        list.addToItems(new StatefulSetBuilder()
-                .withNewMetadata()
-                .withName(name)
-                .endMetadata()
-                .withNewSpec()
-                .withReplicas(1)
-                .withServiceName(name)
-                .withNewSelector()
-                .withMatchLabels(new HashMap<String, String>())
+        StatefulSetBuilder builder = list.getItems().stream()
+                .filter(i -> Strings.isNotNullOrEmpty(i.getKind()) && i.getKind().equalsIgnoreCase(STATEFULSET))
+                .filter(i -> i.getMetadata().getName().equals(name))
+                .map(metadata -> {
+                    list.removeFromItems(metadata);
+                    return new StatefulSetBuilder((StatefulSet) metadata);
+                }).findAny()
+                .orElseGet(() -> new StatefulSetBuilder()
+                        .withNewMetadata()
+                        .withName(name)
+                        .endMetadata());
+
+        // init main sub items
+        StatefulSetFluent.SpecNested<StatefulSetBuilder> specBuilder = builder.editOrNewSpec()
```

Comment:
This is not a change request but more of a suggestion:

```java
        StatefulSet statefulSet = list.getItems().stream()
                .filter(i -> Strings.isNotNullOrEmpty(i.getKind()) && i.getKind().equalsIgnoreCase(STATEFULSET))
                .filter(i -> i.getMetadata().getName().equals(name))
                .map(metadata -> {
                    list.removeFromItems(metadata);
                    return new StatefulSetBuilder((StatefulSet) metadata);
                }).findAny()
                .orElseGet(() -> new StatefulSetBuilder()
                        .withNewMetadata()
                        .withName(name)
                        .endMetadata())
               .accept(StatefulSetSpecBuilder.class, spec -> {
                        if (spec.getReplicas() != null) {
                            spec.withReplicas(1);
                        }
                        if (Strings.isNullOrEmpty(spec.getServiceName())) {
                            spec.withServiceName(name);
                        }
                        if (spec.getSelector().getMatchLabels() == null) {
                            spec.editSelector().withMatchLabels(new HashMap<>()).endSelector();
                        }
                        // - termination grace period seconds
                        if (spec.getTemplate().getSpec().getTerminationGracePeriodSeconds() == null) {
                            spec.editTemplate().editSpec().withTerminationGracePeriodSeconds(10L).endSpec().endTemplate();
                        }
                        //containers
                        if (!containsContainerWithName(spec)) {
                           spec.editTemplate()
                                  .editSpec()
                                  .addNewContainer()
                                  .withName(name)
                                  .endContainer()
                                  .endSpec().endTemplate();
                            } 
                         }).build(); 
```

This mostly avoid referring to internal / nested builder types and gives a more fluent feel.

Reply:
I've refactored this method according to your suggestion.